### PR TITLE
Extending GB for modules to non-commutative 

### PR DESCRIPTION
**Pull request by Lax202**
This is a draft for extending Groebner basis machinery for submodules to non-commutative rings. This is to support the pull request [#3900](https://github.com/oscar-system/Oscar.jl/pull/3900) in Oscar.jl

### CODE REVIEW

File: src/module/ModuleTypes.jl [hunk 1]
```diff
@@ -4,21 +4,27 @@
 #
 ###############################################################################
 
-const FreeModID = Dict{Tuple{PolyRing, Int}, Module}()
+const FreeModID = Dict{Tuple{Union{PolyRing,PluralRing}, Int}, Module}()
 
-mutable struct FreeMod{T <: Nemo.RingElem} <: Module{T}
-   base_ring::PolyRing
+mutable struct FreeMod{T <: Union{Nemo.RingElem, Singular.spluralg}} <: Module{T}
```
Comment (fingolfin) [inline src/module/ModuleTypes.jl:9]: How about just changing this to 
```suggestion
mutable struct FreeMod{T <: NCRingElem} <: Module{T}
```
And similar in the definition of `FreeModID`, etc. etc.?

That way we are ready for supporting other rings if we need to, and it doesn't really cost us much if anything.

File: src/module/module.jl [hunk 2]
```diff
@@ -457,11 +457,21 @@ function Module(R::PolyRing{T}, vecs::svector{spoly{T}}...) where T <: Nemo.Ring
...
+   S = elem_type(R)
+   return smodule{S}(R, vecs...)
+end
+
 function Module(R::PolyRing{T}, id::libSingular.ideal_ptr) where T <: Nemo.RingElem
    S = elem_type(R)
    return smodule{S}(R, id)
 end
 
+function Module(R::PluralRing, id::libSingular.ideal_ptr)
```
Comment (fingolfin) [inline src/module/module.jl:470]: This is exact same as the previous method, just with the `T` dropped, isn't it? So then you can remove the previous method? Or keep it, change the `Nemo.RingElem` to `NCRingElem` (I think the `Nemo.` prefix can be dropped these days, too), and use that?

File: src/module/ModuleTypes.jl [hunk 3]
```diff
@@ -31,9 +31,9 @@ mutable struct svector{T <: Nemo.RingElem} <: Nemo.ModuleElem{T}
    end
 end
 
+
 """
     (R::PolyRing{T})(m::libSingular.poly,::Val{:vector}) where T
-
```
Comment (fingolfin) [inline src/module/ModuleTypes.jl:36]: There really should be an empty line here.

File: src/module/module.jl [hunk 4]
```diff
@@ -38,7 +38,7 @@ function checkbounds(I::smodule, i::Int)
    (i > ngens(I) || i < 1) && throw(BoundsError(I, i))
 end
 
-function getindex(I::smodule{T}, i::Int) where T <: AbstractAlgebra.RingElem
+function getindex(I::smodule{T}, i::Int) where T
```
Comment (fingolfin) [inline src/module/module.jl:41]: Maybe
```suggestion
function getindex(I::smodule{T}, i::Int) where T <: Nemo.NCRingElem
```

File: src/module/vector.jl [hunk 5]
```diff
@@ -138,17 +138,33 @@ function *(a::svector{spoly{T}}, b::spoly{T}) where T <: AbstractAlgebra.RingEle
    return svector{spoly{T}}(R, a.rank, s)
 end
 
+function *(a::svector{Singular.spluralg{T}}, b::Singular.spluralg{T}) where T <: AbstractAlgebra.RingElem
+   base_ring(a) != parent(b) && error("Incompatible base rings")
+   R = base_ring(a)
+   s = GC.@preserve a b R libSingular.pp_Mult_qq(a.ptr, b.ptr, R.ptr)
```
Comment (fingolfin) [inline src/module/vector.jl:144]: Is `libSingular.pp_Mult_qq` the correct function to call here? Maybe @hannes14 can comment

File: src/module/vector.jl [hunk 6]
```diff
@@ -138,17 +138,33 @@ function *(a::svector{spoly{T}}, b::spoly{T}) where T <: AbstractAlgebra.RingEle
...
    base_ring(b) != parent(a) && error("Incompatible base rings")
    R = base_ring(b)
    s = GC.@preserve a b R libSingular.pp_Mult_qq(a.ptr, b.ptr, R.ptr)
    return svector{spoly{T}}(R, b.rank, s)
 end
 
+function *(b::Singular.spluralg{T}, a::svector{Singular.spluralg{T}}) where T
+   base_ring(a) != parent(b) && error("Incompatible base rings")
+   R = base_ring(a)
+   s = GC.@preserve a b R libSingular.pp_Mult_qq(a.ptr, b.ptr, R.ptr)
```
Comment (fingolfin) [inline src/module/vector.jl:158]: Same question here: is this the right function to call here?

Comment (hannes14) [reply src/module/vector.jl:158]: Yes, pp_Mult_qq (and most other p_*,pp_* routines) work for all types of Singular rings.

**Pull request merged**